File: code/packages/go/rom-bios/rom.go
package rombios

// === ROM: Read-Only Memory ===
//
// Real computers have a ROM chip soldered to the motherboard containing
// firmware that executes on power-on. The CPU's program counter starts at
// the ROM's base address (typically 0xFFFF0000 for our simulated machine).
//
// ROM has a distinctive property: writes are silently ignored. The data
// was "burned in" when the ROM was manufactured (or in our case, when the
// ROM object is created). This is like a recipe card sealed in plastic --
// you can read it any number of times, but you cannot modify it.
//
//     Memory map showing ROM's position:
//
//     0xFFFF_FFFF ┌──────────────────┐
//                 │    ROM (64 KB)   │ ← CPU starts executing here
//     0xFFFF_0000 ├──────────────────┤
//                 │   Framebuffer    │
//     0xFFFB_0000 ├──────────────────┤
//                 │       ...        │
//     0x0001_0000 ├──────────────────┤
//                 │   Bootloader     │ ← BIOS jumps here after init
//     0x0000_1000 ├──────────────────┤
//                 │   HardwareInfo   │ ← BIOS writes hardware config here
//     0x0000_0000 ├──────────────────┤
//                 │       IDT        │ ← Interrupt Descriptor Table
//                 └──────────────────┘

// DefaultROMBase is the standard base address for ROM: the top of the
// 32-bit address space minus 64 KB. This is where the CPU's program
// counter points on power-on.
const DefaultROMBase uint32 = 0xFFFF0000

// DefaultROMSize is the standard ROM size: 64 KB (65536 bytes).
// This is enough space for the BIOS firmware.
const DefaultROMSize = 65536

// ROMConfig defines the read-only memory region.
//
// The default configuration places ROM at the top of the 32-bit address
// space (0xFFFF0000), which is the conventional reset vector. The CPU's
// program counter is set to this address on power-on, so whatever code
// lives here executes first.
type ROMConfig struct {
	BaseAddress uint32 // Start address (default: 0xFFFF0000)
	Size        int    // Size in bytes (default: 65536 = 64KB)
}

// DefaultROMConfig returns the standard ROM configuration:
//
//	BaseAddress: 0xFFFF0000, Size: 65536 (64KB).
func DefaultROMConfig() ROMConfig {
	return ROMConfig{
		BaseAddress: DefaultROMBase,
		Size:        DefaultROMSize,
	}
}

// ROM represents a read-only memory region.
//
// Once created with a firmware image, the contents cannot be changed.
// Write operations are silently ignored -- this models the behavior of
// real ROM chips, which are programmed at the factory and cannot be
// rewritten by the CPU.
//
// Example usage:
//
//	bios := NewBIOSFirmware(DefaultBIOSConfig())
//	rom := NewROM(DefaultROMConfig(), bios.Generate())
//
//	// Reading works normally:
//	firstByte := rom.Read(0xFFFF0000)
//	firstWord := rom.ReadWord(0xFFFF0000)
//
//	// Writing is silently ignored:
//	rom.Write(0xFFFF0000, 0xFF)
//	// rom.Read(0xFFFF0000) still returns the original byte
type ROM struct {
	config ROMConfig
	data   []byte
}

// NewROM creates a ROM loaded with the given firmware bytes.
//
// The firmware is copied into internal storage -- the caller's slice
// is not retained. If len(firmware) < config.Size, the remaining
// bytes are zero-filled. If len(firmware) > config.Size, it panics.
//
// Example:
//
//	firmware := []byte{0x37, 0x12, 0x00, 0x00} // lui x4, 0x12
//	rom := NewROM(DefaultROMConfig(), firmware)
func NewROM(config ROMConfig, firmware []byte) *ROM {
	if len(firmware) > config.Size {
		panic("firmware larger than ROM size")
	}

	// Copy firmware into a fixed-size buffer (zero-filled beyond firmware)
	data := make([]byte, config.Size)
	copy(data, firmware)

	return &ROM{
		config: config,
		data:   data,
	}
}

// Read returns a single byte from the given absolute address.
//
// The address must fall within [BaseAddress, BaseAddress+Size).
// Out-of-range addresses return 0, modeling the behavior of reading
// from unmapped memory.
func (r *ROM) Read(address uint32) byte {
	offset := r.addressToOffset(address)
	if offset < 0 {
		return 0
	}
	return r.data[offset]
}

// ReadWord returns a 32-bit little-endian word starting at the given
// absolute address.
//
// This is the primary access pattern since RISC-V instructions are
// 32 bits wide. The CPU fetches one word at a time from ROM.
//
// Little-endian means the least significant byte is at the lowest address:
//
//	Address+0: byte 0 (bits 7:0)
//	Address+1: byte 1 (bits 15:8)
//	Address+2: byte 2 (bits 23:16)
//	Address+3: byte 3 (bits 31:24)
func (r *ROM) ReadWord(address uint32) uint32 {
	offset := r.addressToOffset(address)
	if offset < 0 || offset+3 >= len(r.data) {
		return 0
	}
	return readLE32(r.data[offset:])
}

// Write attempts to write a byte to ROM. Since ROM is read-only,
// this operation is silently ignored.
//
// In a real system, attempting to write to ROM has no effect. The
// data lines simply do not connect to write circuitry. We model this
// by accepting the call but doing nothing.
func (r *ROM) Write(address uint32, value byte) {
	// Silently ignored -- ROM is read-only.
	_ = address
	_ = value
}

// Size returns the total size of the ROM in bytes.
func (r *ROM) Size() int {
	return r.config.Size
}

// BaseAddress returns the base address of the ROM.
func (r *ROM) BaseAddress() uint32 {
	return r.config.BaseAddress
}

// Contains returns true if the given address falls within the ROM region.
func (r *ROM) Contains(address uint32) bool {
	return r.addressToOffset(address) >= 0
}

// addressToOffset converts an absolute address to a byte offset within
// the ROM's data buffer. Returns -1 if the address is out of range.
func (r *ROM) addressToOffset(address uint32) int {
	if address < r.config.BaseAddress {
		return -1
	}
	offset := int(address - r.config.BaseAddress)
	if offset >= r.config.Size {
		return -1
	}
	return offset
}

File: code/packages/rust/rom-bios/src/rom.rs
//! ROM (Read-Only Memory) implementation.
//!
//! ROM is a memory region where writes are silently ignored. Real computers
//! have a ROM chip containing firmware that the CPU executes on power-on.
//! The program counter starts at the ROM's base address (0xFFFF0000).
//!
//! Analogy: ROM is like a recipe card laminated in plastic -- you can read
//! it any number of times, but you cannot write on it.

/// Default base address: top of 32-bit address space minus 64 KB.
pub const DEFAULT_ROM_BASE: u32 = 0xFFFF0000;

/// Default ROM size: 64 KB (65536 bytes).
pub const DEFAULT_ROM_SIZE: usize = 65536;

/// Configuration for the ROM memory region.
#[derive(Debug, Clone)]
pub struct RomConfig {
    pub base_address: u32,
    pub size: usize,
}

impl Default for RomConfig {
    fn default() -> Self {
        Self {
            base_address: DEFAULT_ROM_BASE,
            size: DEFAULT_ROM_SIZE,
        }
    }
}

/// Read-only memory region. Writes are silently ignored.
///
/// # Example
/// ```
/// use rom_bios::rom::{Rom, RomConfig};
///
/// let rom = Rom::new(RomConfig::default(), &[0xAA, 0xBB, 0xCC, 0xDD]);
/// assert_eq!(rom.read(0xFFFF0000), 0xAA);
/// ```
pub struct Rom {
    config: RomConfig,
    data: Vec<u8>,
}

impl Rom {
    /// Create a ROM loaded with the given firmware bytes.
    ///
    /// If firmware is shorter than config.size, remaining bytes are zero-filled.
    /// Panics if firmware is larger than config.size.
    pub fn new(config: RomConfig, firmware: &[u8]) -> Self {
        assert!(firmware.len() <= config.size, "firmware larger than ROM size");
        let mut data = vec![0u8; config.size];
        data[..firmware.len()].copy_from_slice(firmware);
        Self { config, data }
    }

    /// Read a single byte from the given absolute address.
    /// Out-of-range addresses return 0.
    pub fn read(&self, address: u32) -> u8 {
        match self.address_to_offset(address) {
            Some(offset) => self.data[offset],
            None => 0,
        }
    }

    /// Read a 32-bit little-endian word at the given absolute address.
    pub fn read_word(&self, address: u32) -> u32 {
        match self.address_to_offset(address) {
            Some(offset) if offset + 3 < self.data.len() => {
                u32::from_le_bytes([
                    self.data[offset],
                    self.data[offset + 1],
                    self.data[offset + 2],
                    self.data[offset + 3],
                ])
            }
            _ => 0,
        }
    }

    /// Attempt to write a byte to ROM (silently ignored).
    pub fn write(&self, _address: u32, _value: u8) {
        // ROM is read-only. Silently ignored.
    }

    /// Total size of ROM in bytes.
    pub fn size(&self) -> usize {
        self.config.size
    }

    /// Base address of ROM.
    pub fn base_address(&self) -> u32 {
        self.config.base_address
    }

    /// True if address falls within the ROM region.
    pub fn contains(&self, address: u32) -> bool {
        self.address_to_offset(address).is_some()
    }

    fn address_to_offset(&self, address: u32) -> Option<usize> {
        if address < self.config.base_address {
            return None;
        }
        let offset = (address - self.config.base_address) as usize;
        if offset >= self.config.size {
            return None;
        }
        Some(offset)
    }
}
